mono/metadata/marshal-ilgen.c

```diff
@@ -6483,8 +6483,21 @@ emit_native_icall_wrapper_ilgen (MonoMethodBuilder *mb, MonoMethod *method, Mono
 		}
 		mono_mb_emit_ldloc_addr (mb, error_var);
 	} else {
-		for (int i = 0; i < csig->param_count; i++)
+		for (int i = 0; i < csig->param_count; i++) {
 			mono_mb_emit_ldarg (mb, i);
+
+			// Make parameters volatile to pin them.
+
+			if (get_method_image (method) && ((i == 0 && csig->hasthis) || MONO_TYPE_IS_REFERENCE (csig->params [i]) || mono_type_is_byref (csig->params [i]))) {
+				if (!mb->volatile_args) {
+					gpointer const mem = mono_image_alloc0 (get_method_image (method), mono_bitset_alloc_size (csig->param_count, 0));
```

Comment:
Probably should just use another allocator if there is no image. Like for dynamic? Like for Tasklets?

Reply:
But at least this should cover the runtime/mscorlib.dll.